### DISCUSSION TITLE
Scatter plot with fixed markersize and hue no longer possible since v2022.10 ?

### DISCUSSION THREAD

**Discussion by jorisvandenbossche** [pydata/xarray#7268] [General]
With v2022.09 and before, the scatter() plot method had the following behaviour:

```python
import xarray as xr
temp = xr.DataArray(np.random.randn(10, 10, 2), coords=[np.arange(10), np.arange(10), [2021, 2022]], dims=["lon", "lat", "year"])
prec = xr.DataArray(np.random.randn(10, 10, 2), coords=[np.arange(10), np.arange(10), [2021, 2022]], dims=["lon", "lat", "year"])
ds = xr.Dataset({"temperature": temp, "precipitation": prec})

>>> ds
<xarray.Dataset>
Dimensions:        (lon: 10, lat: 10, year: 2)
Coordinates:
  * lon            (lon) int64 0 1 2 3 4 5 6 7 8 9
  * lat            (lat) int64 0 1 2 3 4 5 6 7 8 9
  * year           (year) int64 2021 2022
Data variables:
    temperature    (lon, lat, year) float64 0.5505 -0.2942 ... 1.085 2.287
    precipitation  (lon, lat, year) float64 0.6686 0.2782 ... 1.073 -1.053

>>> ds.plot.scatter(x="temperature", y="precipitation")
```

![test_xarray](https://user-images.githubusercontent.com/1020496/200603646-b57a71ee-48f7-40f9-b9c0-307e4796e1c5.png)

So creating a simple scatter plot for all points (across dimensions) for the two variables. 

With the latest xarray, we now get a different result for the same code:

![test_xarray2](https://user-images.githubusercontent.com/1020496/200604139-7f39474a-3097-44f6-a302-bd68f5dd63fb.png)

So here xarray somehow decided to use the "year" dimension for markersize, and "lat" for color (hue) (I suppose starting from the last variable / dimension not used for x/y). 

This is already a (breaking, but potentially intentional) change in behaviour, but I also don't seem to be able to get the previous behaviour to use a fixed color and markersize. It seems that both keywords only accept a dimension name, and not a constant value:

```python
>>> ds.plot.scatter(x="temperature", y="precipitation", markersize=1)
...
ValueError: size must be one of None, 'lat', 'lon', 'temperature', 'year'. Received '1' instead.

>>> ds.plot.scatter(x="temperature", y="precipitation", hue="blue")
...
ValueError: hue must be one of None, 'lat', 'lon', 'temperature', 'year'. Received 'blue' instead.

# no error, but also has no effect on actual colors (same result as above without specifying c)
>>> ds.plot.scatter(x="temperature", y="precipitation", c="blue")
<matplotlib.collections.PathCollection at 0x7f55b7109990>
```

Also specifying eg `hue=None` to explicitly say _"don't use a variable to control the hue"_ doesn't help.

---

Looking at the whatsnew notes, this might be related to https://github.com/pydata/xarray/pull/6778? (didn't check though) 

And assuming the change in behaviour was intentional, is it also on purpose there is no way to _not_ use a variable markersizer or hue?

**Comment by Illviljan** [comment]
To make a guess which dimension to use is intentional, as there has been a precedent of guessing dimensions in other plotting methods:

```python
import numpy as np
import matplotlib.pyplot as plt

import xarray as xr

temp = xr.DataArray(np.random.randn(10, 10, 2), coords=[np.arange(10), np.arange(10), [2021, 2022]], dims=["lon", "lat", "year"])
prec = xr.DataArray(np.random.randn(10, 10, 2), coords=[np.arange(10), np.arange(10), [2021, 2022]], dims=["lon", "lat", "year"])
one = xr.DataArray(np.random.randn(10, 10, 2), coords=[np.arange(10), np.arange(10), [2021, 2022]], dims=["lon", "lat", "year"])
blue = xr.DataArray(np.random.randn(10, 10, 2), coords=[np.arange(10), np.arange(10), [2021, 2022]], dims=["lon", "lat", "year"])
ds = xr.Dataset({"temperature": temp, "precipitation": prec, 1: one, "blue": blue})

# line adds hue:
ds.temperature.sel(lat=0).plot.line(x="lon")

# contourf adds hue:
ds.temperature.sel(year=2021).plot.contourf()
```
And I believe that `.scatter` simply didn't do this before because it was a little tricky getting categorical and numerical data to work nicely together.

You can do one of these if information about dims is of no interest:

```python
# Stack the non interesting dims:
ds.stack(stacked_dim=[...]).plot.scatter(x="temperature", y="precipitation")

# Use matplotlib:
import matplotilib.pyplot as plt

fig, ax = plt.subplots()
ax.scatter(ds.temperature, ds.precipitation)
```

`hue=None` is the default value. If hue is none, we try to guess a value:
https://github.com/pydata/xarray/blob/cc7e09a3507fa342b3790b5c109e700fa12f0b17/xarray/plot/dataarray_plot.py#L154
It's tricky to choose another default that can imply "Do not plot" however. Because xarray allows all hashables as a key.

```python
# 1 and "blue" is hashable and maybe someone wants to use that as an array name:
fig, ax = plt.subplots()
ds.plot.scatter(x="temperature", y="precipitation", hue="blue")
```

**Reply by jorisvandenbossche** [reply]
I understand that the choice was made to change the default behaviour (and that this made it more consistent with other plotting methods), but I do want to say that I find it a regression in usability that it is no longer possible to specify a constant value for one of the plotting characteristics (color, size, marker) in case you case you have more than 2 dimensions in your dataset but which are not suitable to be used for those plotting characteristics (without manually having to stack those dimensions).

>  It's tricky to choose another default that can imply "Do not plot" however. Because xarray allows all hashables as a key.

Yes, I certainly understand, and I didn't want to suggest that this would be the best way to reuse the same keyword. I was just trying the different possibilities that might have worked. 

As a reference, I checked seaborn, and they do allow to specify a fixed color and size using matplotlib's keywords for this (`c`/`color` and `s` in case of a scatter plot). 
For an example, see one of the example towards the bottom of the scatterplot reference page (https://seaborn.pydata.org/generated/seaborn.scatterplot.html), which says:

> Additional keyword arguments are passed to [matplotlib.axes.Axes.scatter()](https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.scatter.html#matplotlib.axes.Axes.scatter), allowing you to directly set the attributes of the plot that are not semantically mapped:
>
>     sns.scatterplot(data=tips, x="total_bill", y="tip", s=100, color=".2", marker="+")

This could be something that xarray could maybe also allow? It could be possible to check if those are in `kwargs` and in that case not guess which dimension to use for `hue` and `markersize` and not map those to `c` and `s`, but instead use the user-specified value?

**Reply by mgunyho** [reply]
> It's tricky to choose another default that can imply "Do not plot" however

I have used at least one library where `None` was used as the "default" and `False` was used for "don't do anything". Perhaps that could work here?

**Comment by dcherian** [comment]
Yeah I ran in to this too. I think we should revert the auto-guessing bit.

> there has been a precedent of guessing dimensions in other plotting methods:

This precedent only applies when its unambiguous. Otherwise we error.

We should only ever make unambiguous choices, and force the user to be explicit about what they want otherwise

**Comment by stijnvanhoey** [comment]
I'm not sure if I should raise a different issue, but this seems directly linked to the fixed versus auto-guessing of `hue` with the scatter method:

```python
>>> ds = xr.Dataset(data_vars=dict(pressure=(("level", "date"), np.arange(60).reshape(-1, 30)),
...                                temperature=(("level", "date"), np.random.randn(60).reshape(-1, 30))
...                               ),
...                 coords=dict(date=pd.date_range(start="2022-01-01", periods=30, freq="D"),
...                             level=np.arange(2)
...                            )
...                )
>>> ds
<xarray.Dataset>
Dimensions:      (level: 2, date: 30)
Coordinates:
  * date         (date) datetime64[ns] 2022-01-01 2022-01-02 ... 2022-01-30
  * level        (level) int64 0 1
Data variables:
    pressure     (level, date) int64 0 1 2 3 4 5 6 7 ... 52 53 54 55 56 57 58 59
    temperature  (level, date) float64 1.06 -1.668 0.3954 ... -0.7142 0.9975
>>> ds.plot.scatter(x="date", y="temperature", col="level")
```
returns a colorbar (auto-added with temperature data), but the colors of the scatter markers are not adjusted: 
![image](https://user-images.githubusercontent.com/754862/201754668-7f931bd3-75bb-4297-8c41-1acd4d92443a.png)

As far as I understood, I expect the color of the markers would be mapped by the temperature, i.e. current default behavior by the auto-guessing? Still, I would prefer the color mapping of the markers with another variable only when explicitly defined by the user:

```python
>>> ds.plot.scatter(x="date", y="temperature", col="level")
```
no `hue` mapping defined, no color mapping of the markers

```python
>>> ds.plot.scatter(x="date", y="temperature", col="level", hue="temperature")
```
`hue` mapping defined, color mapping of the markers and colorbar added
